File: library/src/main/java/com/mikepenz/materialdrawer/model/ContainerDrawerItem.java
package com.mikepenz.materialdrawer.model;

import android.content.Context;
import android.support.annotation.LayoutRes;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;

import com.mikepenz.materialdrawer.R;
import com.mikepenz.materialdrawer.holder.DimenHolder;
import com.mikepenz.materialize.util.UIUtils;

import java.util.List;

/**
 * Created by mikepenz on 03.02.15.
 */
public class ContainerDrawerItem extends AbstractDrawerItem<ContainerDrawerItem, ContainerDrawerItem.ViewHolder> {

    private DimenHolder mHeight;

    public ContainerDrawerItem withHeight(DimenHolder height) {
        mHeight = height;
        return this;
    }

    public DimenHolder getHeight() {
        return mHeight;
    }

    private View mView;

    public ContainerDrawerItem withView(View view) {
        this.mView = view;
        return this;
    }

    public View getView() {
        return mView;
    }

    public enum Position {
        TOP,
        BOTTOM,
        NONE;
    }

    private Position mViewPosition = Position.TOP;

    public ContainerDrawerItem withViewPosition(Position position) {
        this.mViewPosition = position;
        return this;
    }

    private boolean mDivider = true;

    public ContainerDrawerItem withDivider(boolean divider) {
        this.mDivider = divider;
        return this;
    }

    public Position getViewPosition() {
        return mViewPosition;
    }

    @Override
    public int getType() {
        return R.id.material_drawer_item_container;
    }

    @Override
    @LayoutRes
    public int getLayoutRes() {
        return R.layout.material_drawer_item_container;
    }

    @Override
    public void bindView(ViewHolder viewHolder, List payloads) {
        super.bindView(viewHolder, payloads);

        Context ctx = viewHolder.itemView.getContext();

        //set the identifier from the drawerItem here. It can be used to run tests
        viewHolder.itemView.setId(hashCode());

        //define how the divider should look like
        viewHolder.view.setEnabled(false);

        //make sure our view is not used in another parent
        if (mView.getParent() != null) {
            ((ViewGroup) mView.getParent()).removeView(mView);
        }

        //set the height
        int height = ViewGroup.LayoutParams.WRAP_CONTENT;
        if (mHeight != null) {
            RecyclerView.LayoutParams lp = (RecyclerView.LayoutParams) viewHolder.view.getLayoutParams();
            height = mHeight.asPixel(ctx);
            lp.height = height;
            viewHolder.view.setLayoutParams(lp);
        }

        //make sure the header view is empty
        ((ViewGroup) viewHolder.view).removeAllViews();

        int dividerHeight = 0;
        if (mDivider) {
            dividerHeight = 1;
        }

        View divider = new View(ctx);
        divider.setMinimumHeight(dividerHeight);
        divider.setBackgroundColor(UIUtils.getThemeColorFromAttrOrRes(ctx, R.attr.material_drawer_divider, R.color.material_drawer_divider));

        LinearLayout.LayoutParams dividerParams = new LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, (int) UIUtils.convertDpToPixel(dividerHeight, ctx));
        LinearLayout.LayoutParams viewParams = new LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, mHeight != null ? height - (int) UIUtils.convertDpToPixel(dividerHeight, ctx) : height);

        //depending on the position we add the view
        if (mViewPosition == Position.TOP) {
            ((ViewGroup) viewHolder.view).addView(mView, viewParams);
            dividerParams.bottomMargin = ctx.getResources().getDimensionPixelSize(R.dimen.material_drawer_padding);
            ((ViewGroup) viewHolder.view).addView(divider, dividerParams);
        } else if (mViewPosition == Position.BOTTOM) {
            dividerParams.topMargin = ctx.getResources().getDimensionPixelSize(R.dimen.material_drawer_padding);
            ((ViewGroup) viewHolder.view).addView(divider, dividerParams);
            ((ViewGroup) viewHolder.view).addView(mView, viewParams);
        } else {
            ((ViewGroup) viewHolder.view).addView(mView, viewParams);
        }

        //call the onPostBindView method to trigger post bind view actions (like the listener to modify the item if required)
        onPostBindView(this, viewHolder.itemView);
    }

    @Override
    public ViewHolder getViewHolder(View v) {
        return new ViewHolder(v);
    }

    public static class ViewHolder extends RecyclerView.ViewHolder {
        private View view;

        private ViewHolder(View view) {
            super(view);
            this.view = view;
        }
    }
}

File: library/src/main/java/com/mikepenz/materialdrawer/model/ContainerDrawerItem.kt
package com.mikepenz.materialdrawer.model

import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import androidx.annotation.LayoutRes
import androidx.recyclerview.widget.RecyclerView
import com.mikepenz.materialdrawer.R
import com.mikepenz.materialdrawer.holder.DimenHolder
import com.mikepenz.materialize.util.UIUtils

/**
 * Created by mikepenz on 03.02.15.
 */
open class ContainerDrawerItem : AbstractDrawerItem<ContainerDrawerItem, ContainerDrawerItem.ViewHolder>() {

    var height: DimenHolder? = null
    var view: View? = null
    var viewPosition = Position.TOP
    var divider = true

    override val type: Int
        get() = R.id.material_drawer_item_container

    override val layoutRes: Int
        @LayoutRes
        get() = R.layout.material_drawer_item_container

    fun withHeight(height: DimenHolder?): ContainerDrawerItem {
        this.height = height
        return this
    }

    fun withView(view: View): ContainerDrawerItem {
        this.view = view
        return this
    }

    enum class Position {
        TOP,
        BOTTOM,
        NONE
    }

    fun withViewPosition(position: Position): ContainerDrawerItem {
        this.viewPosition = position
        return this
    }

    fun withDivider(_divider: Boolean): ContainerDrawerItem {
        this.divider = _divider
        return this
    }

    override fun bindView(holder: ViewHolder, payloads: MutableList<Any>) {
        super.bindView(holder, payloads)

        val ctx = holder.itemView.context

        //set the identifier from the drawerItem here. It can be used to run tests
        holder.itemView.id = hashCode()

        //define how the divider should look like
        holder.view.isEnabled = false

        //make sure our view is not used in another parent
        view?.parent?.let {
            (it as ViewGroup).removeView(view)
        }

        //set the height
        var height = ViewGroup.LayoutParams.WRAP_CONTENT

        this.height?.let {
            val lp = holder.view.layoutParams as RecyclerView.LayoutParams
            height = it.asPixel(ctx)
            lp.height = height
            holder.view.layoutParams = lp
        }

        //make sure the header view is empty
        (holder.view as ViewGroup).removeAllViews()

        var dividerHeight = 0
        if (divider) {
            dividerHeight = 1
        }

        val divider = View(ctx)
        divider.minimumHeight = dividerHeight
        divider.setBackgroundColor(UIUtils.getThemeColorFromAttrOrRes(ctx, R.attr.material_drawer_divider, R.color.material_drawer_divider))

        val dividerParams = LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, UIUtils.convertDpToPixel(dividerHeight.toFloat(), ctx).toInt())
        val viewParams = LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, if (this.height != null) height - UIUtils.convertDpToPixel(dividerHeight.toFloat(), ctx).toInt() else height)

        //depending on the position we add the view
        when (viewPosition) {
            Position.TOP -> {
                holder.view.addView(view, viewParams)
                dividerParams.bottomMargin = ctx.resources.getDimensionPixelSize(R.dimen.material_drawer_padding)
                holder.view.addView(divider, dividerParams)
            }
            Position.BOTTOM -> {
                dividerParams.topMargin = ctx.resources.getDimensionPixelSize(R.dimen.material_drawer_padding)
                holder.view.addView(divider, dividerParams)
                holder.view.addView(view, viewParams)
            }
            else -> holder.view.addView(view, viewParams)
        }

        //call the onPostBindView method to trigger post bind view actions (like the listener to modify the item if required)
        onPostBindView(this, holder.itemView)
    }

    override fun getViewHolder(v: View): ViewHolder {
        return ViewHolder(v)
    }

    class ViewHolder internal constructor(internal val view: View) : RecyclerView.ViewHolder(view)
}
